File: javascript/freelancer-rates/freelancer-rates.js
// @ts-check

const BILLABLE_HOURS_PER_DAY = 8
const BILLABLE_DAYS_PER_MONTH = 22

/**
 * The day rate, given a rate per hour
 *
 * @param {number} ratePerHour
 * @returns {number} the rate per day
 */
export function dayRate(ratePerHour) {
  return ratePerHour * BILLABLE_HOURS_PER_DAY
}

/**
 * Calculates the number of days in a budget, rounded down
 *
 * @param {number} budget: the total budget
 * @param {number} ratePerHour: the rate per hour
 * @returns {number} the number of days
 */
export function daysInBudget(budget, ratePerHour) {
  return Math.floor(budget / ratePerHour / BILLABLE_HOURS_PER_DAY)
}

/**
 * Calculates the discounted rate for large projects, rounded up
 *
 * @param {number} ratePerHour
 * @param {number} numDays: number of days the project spans
 * @param {number} discount: for example 20% written as 0.2
 * @returns {number} the rounded up discounted rate
 */
export function priceWithMonthlyDiscount(ratePerHour, numDays, discount) {
  const numMonths = Math.floor(numDays / BILLABLE_DAYS_PER_MONTH)
  const totalDiscount = numMonths * BILLABLE_DAYS_PER_MONTH * dayRate(ratePerHour) * discount
  const totalCosts = dayRate(ratePerHour) * numDays
  return Math.ceil(totalCosts - totalDiscount)
}

File: cpp/freelancer-rates/freelancer_rates.cpp
#include <cmath>

// daily_rate calculates the daily rate given an hourly rate
double daily_rate(double hourly_rate)
{
    return hourly_rate * 8;
}

// apply_discount calculates the price after a discount
double apply_discount(double before_discount, double discount)
{
    return before_discount - (discount / 100.0) * before_discount;
}

// monthly_rate calculates the monthly rate, given an hourly rate and a discount
// The returned monthly rate is rounded up to the nearest integer.
int monthly_rate(double hourly_rate, double discount)
{
    return ceil(apply_discount(daily_rate(hourly_rate) * 22, discount));
}

// days_in_budget calculates the number of workdays given a budget, hourly rate,
// and discount The returned number of days is rounded down (take the floor) to
// the next integer.
int days_in_budget(int budget, double hourly_rate, double discount)
{
    return floor(budget / (apply_discount(hourly_rate, discount) * 8));
}